### PR TITLE
message queue tests, requirements and traceability

### PR DESCRIPTION
**Pull request by darobs**
I needed to create this queue to prevent a dead remote from blocking the gateway. The Outprocess_Receive() function now pushes the message it receives on the queue, and an outgoing messages thread sends messages in FIFO order to the remote.  The prevents the gateway from blocking until memory runs out.

Unfortunately, I commited the implementation on the feature branch first, if you want to see that, use this link:
https://github.com/darobs/azure-iot-gateway-sdk/compare/0d2799289b3658f7517f394c08e2ac1f090cc528...darobs:oop-modules-message-queue

### CODE REVIEW

**Comment by aribeironovaes** [comment]
My First comment (overall comment), this PR could go to our Main Repo instead of your Fork (I can think of a way of doing that with 2 branches. Why? so it get public visibility... PR here will get lost in your repo... 

Just an idea.

File: core/devdoc/message_queue_requirements.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+MESSAGE QUEUE REQUIREMENTS
+==========================
+
+Overview
+--------
+
+The message queue is a very simple queue intended to manage messages. The message queue is typed with MESSAGE_HANDLE because the destruction of the queue requires the destruction of the messages inside the queue.  **Unless the queue is destroyed, the user of this queue is expected to clone before pushing onto the queue, and is expected to destroy the message after popping the message off teh queue.**
```
Comment (aribeironovaes) [inline core/devdoc/message_queue_requirements.md]: >teh [](start = 391, length = 3)

typo.

Comment (aribeironovaes) [inline core/devdoc/message_queue_requirements.md]: > **Unless  [](start = 213, length = 10)

Suggestion putting a new line here so this goes into the next paragraph.

File: core/devdoc/message_queue_requirements.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+----------
+
+[Message requirements](message_requirements.md)
+
+Exposed API
+-----------
+
+```c
+/* creation */
+MESSAGE_QUEUE_HANDLE MESSAGE_QUEUE_create());
```
Comment (aribeironovaes) [inline core/devdoc/message_queue_requirements.md]: >e()); [](start = 40, length = 5)

doesn't compile with an extra ) :)

File: core/devdoc/message_queue_requirements.md (hunk 1 above)
Comment (aribeironovaes) [inline core/devdoc/message_queue_requirements.md]: >The message queue is a very simple queue intended to manage messages. The message queue is typed with MESSAGE [](start = 0, length = 109)

Sorry for this comment. But do we really need this? Why don't we use one of the existing lists that we have already on CShared Utility?

File: core/devdoc/message_queue_requirements.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+Inserts a mesage handle into the message queue.
+
+**SRS_MESSAGE_QUEUE_17_007: [** MESSAGE\_QUEUE\_push shall return a non-zero value if `handle` or `element` are `NULL`. **]**
+
+**SRS_MESSAGE_QUEUE_17_008: [** MESSAGE\_QUEUE\_push shall return zero on success. **]**
+
+**SRS_MESSAGE_QUEUE_17_009: [** MESSAGE\_QUEUE\_push shall return a non-zero value if any system call fails. **]**
+
+**SRS_MESSAGE_QUEUE_17_010: [** A successful call to MESSAGE\_QUEUE\_push on an empty queue will cause the message queue to be non-empty (MESSAGE\_QUEUE\_is\_empty shall return false). **]**
+
```
Comment (aribeironovaes) [inline core/devdoc/message_queue_requirements.md]: I would change this requirement or remove it completely. Requirement 11 is sufficient IMO to make sure the Queue is not empty....

File: core/inc/message_queue.h [hunk 4]
```diff
@@ -23,6 +23,7 @@ typedef struct MESSAGE_QUEUE_TAG* MESSAGE_QUEUE_HANDLE;
 
 /* creation */
 MOCKABLE_FUNCTION(, MESSAGE_QUEUE_HANDLE, MESSAGE_QUEUE_create);
+/* destruction */
```
Comment (aribeironovaes) [inline core/inc/message_queue.h]: Very minor. Adding a blank line to follow the same pattern? :)

File: core/src/message_queue.c [hunk 5]
```diff
@@ -24,15 +24,20 @@ typedef struct MESSAGE_QUEUE_TAG
 static MESSAGE_HANDLE message_pop(MESSAGE_QUEUE_HANDLE_DATA* handle)
 {
     MESSAGE_HANDLE result;
-    MESSAGE_QUEUE_STORAGE* entry = 
+	if (DList_IsListEmpty((PDLIST_ENTRY)&(handle->queue_head)))
```
Comment (aribeironovaes) [inline core/src/message_queue.c]: >if (DList_IsListEmpty((PDLIST_ENTRY)&(handle->queue_head))) [](start = 1, length = 59)

Don't you need to test first if handle is null before dereferencing? (Do you have a test for message_pop where handle is null?

File: core/src/message_queue.c [hunk 6]
```diff
@@ -24,15 +24,20 @@ typedef struct MESSAGE_QUEUE_TAG
...
+		MESSAGE_QUEUE_STORAGE* entry =
 		(MESSAGE_QUEUE_STORAGE*)DList_RemoveHeadList( (PDLIST_ENTRY)&(handle->queue_head));
-    if (entry == &(handle->queue_head))
-    {
-        result = NULL;
-    }
-    else
-    {
+
         result = ((MESSAGE_QUEUE_STORAGE*)entry)->message;
```
Comment (aribeironovaes) [inline core/src/message_queue.c]: >result = ((MESSAGE_QUEUE_STORAGE*)entry)->message; [](start = 8, length = 50)

Don't you need to test if entry is null? Can DList_RemoveHeadList fail?

File: core/src/message_queue.c [hunk 7]
```diff
@@ -68,8 +77,10 @@ void MESSAGE_QUEUE_destroy(MESSAGE_QUEUE_HANDLE handle)
         MESSAGE_HANDLE message;
         while((message = message_pop(mq)) != NULL)
```
Comment (aribeironovaes) [inline core/src/message_queue.c]: >message_pop(mq) [](start = 25, length = 15)

Should you use: MESSAGE_QUEUE_pop instead of directly using message_pop?

File: core/src/message_queue.c [hunk 8]
```diff
@@ -82,23 +93,27 @@ int MESSAGE_QUEUE_push(MESSAGE_QUEUE_HANDLE handle, MESSAGE_HANDLE element)
...
             result = __LINE__;
         }
         else
         {
             DList_InitializeListHead((PDLIST_ENTRY)temp);
             temp->message = element;
+            /*Codes_SRS_MESSAGE_QUEUE_17_010: [ A successful call to MESSAGE_QUEUE_push on an empty queue will cause the message queue to be non-empty (MESSAGE_QUEUE_is_empty shall return false). ]*/
+            /*Codes_SRS_MESSAGE_QUEUE_17_011: [ Messages shall be pushed into the queue in a first-in-first-out order. ]*/
             DList_AppendTailList((PDLIST_ENTRY)&(handle->queue_head), (PDLIST_ENTRY)temp);
+            /*Codes_SRS_MESSAGE_QUEUE_17_008: [ MESSAGE_QUEUE_push shall return zero on success. ]*/
```
Comment (aribeironovaes) [inline core/src/message_queue.c]: Can this fail?

**Comment by aribeironovaes** [comment]
:shipit:

File: core/devdoc/message_queue_requirements.md (hunk 1 above)
Comment (darobs) [reply core/devdoc/message_queue_requirements.md]: fixed

Comment (darobs) [reply core/devdoc/message_queue_requirements.md]: That works.  Done.

File: core/devdoc/message_queue_requirements.md (hunk 2 above)
Comment (darobs) [reply core/devdoc/message_queue_requirements.md]: That's what I get for stripping off all the MOCKABLE_FUNCTION macros. Fixed

File: core/devdoc/message_queue_requirements.md (hunk 1 above)
Comment (darobs) [reply core/devdoc/message_queue_requirements.md]: I admit, I wanted a more easily mockable queue so that I didn't need to worry about the out of process module unit tests being even longer and more complicated than they already are.  This did help, but this unit is pretty basic.  I'm gonna keep it for now, with an eye to remove it if I need to make changes to how the queuing works in the out of process module.

File: core/devdoc/message_queue_requirements.md (hunk 3 above)
Comment (darobs) [reply core/devdoc/message_queue_requirements.md]: Removed.

File: core/inc/message_queue.h (hunk 4 above)
Comment (darobs) [reply core/inc/message_queue.h]: YEs, looks better that way. Done.

File: core/src/message_queue.c (hunk 5 above)
Comment (darobs) [reply core/src/message_queue.c]: This function is internal, so the queue gets a null check before use in \_destroy and \_pop.

File: core/src/message_queue.c (hunk 7 above)
Comment (darobs) [reply core/src/message_queue.c]: No. :)  MESSAGE_QUEUE_pop is the external function, message_pop is the static function that's used internally.

File: core/src/message_queue.c (hunk 8 above)
Comment (darobs) [reply core/src/message_queue.c]: Yes and no. It returns void so I don't have a return value to check. Obviously, it's manipulating pointers, so something could go wrong, but this unit is the owner of the queue, so it shouldn't go wrong.

File: core/src/message_queue.c (hunk 6 above)
Comment (darobs) [reply core/src/message_queue.c]: DList_RemoveHeadList returns the list head if it's empty, so it should never return NULL.
